File: gradient-boosting/GradientBoosting.kt
import kotlin.math.*

data class Tree(
    val prediction: Double? = null,
    val feature: Int? = null,
    val threshold: Double? = null,
    val leftPred: Double? = null,
    val rightPred: Double? = null
)

class GradientBoosting(
    private val nEstimators: Int = 10,
    private val learningRate: Double = 0.1,
    private val maxDepth: Int = 3
) {
    private val trees = mutableListOf<Tree>()
    private var initialPrediction: Double = 0.0
    
    private fun sigmoid(x: Double): Double {
        return 1.0 / (1.0 + exp(-x.coerceIn(-500.0, 500.0)))
    }
    
    fun fit(X: Array<DoubleArray>, y: IntArray) {
        // Initialize with log odds
        val meanY = y.average()
        initialPrediction = ln(meanY / (1 - meanY))
        
        // Current predictions
        val currentPredictions = DoubleArray(y.size) { initialPrediction }
        
        repeat(nEstimators) {
            // Calculate residuals (negative gradient)
            val probabilities = currentPredictions.map { sigmoid(it) }
            val residuals = y.mapIndexed { i, yi -> yi - probabilities[i] }.toDoubleArray()
            
            // Fit a simple regression tree to residuals
            val tree = fitTree(X, residuals)
            trees.add(tree)
            
            // Update predictions
            val treePredictions = predictTree(X, tree)
            for (j in currentPredictions.indices) {
                currentPredictions[j] += learningRate * treePredictions[j]
            }
        }
    }
    
    private fun fitTree(X: Array<DoubleArray>, y: DoubleArray): Tree {
        var bestMse = Double.POSITIVE_INFINITY
        var bestFeature: Int? = null
        var bestThreshold: Double? = null
        
        for (feature in X[0].indices) {
            val thresholds = X.map { it[feature] }.distinct()
            for (threshold in thresholds) {
                val leftMask = X.map { it[feature] <= threshold }
                val yLeft = y.filterIndexed { i, _ -> leftMask[i] }
                val yRight = y.filterIndexed { i, _ -> !leftMask[i] }
                
                if (yLeft.isEmpty() || yRight.isEmpty()) continue
                
                val leftPred = yLeft.average()
                val rightPred = yRight.average()
                
                val mse = yLeft.sumOf { (it - leftPred).pow(2) } + 
                         yRight.sumOf { (it - rightPred).pow(2) }
                
                if (mse < bestMse) {
                    bestMse = mse
                    bestFeature = feature
                    bestThreshold = threshold
                }
            }
        }
        
        if (bestFeature == null) {
            return Tree(prediction = y.average())
        }
        
        val leftMask = X.map { it[bestFeature] <= bestThreshold!! }
        val yLeft = y.filterIndexed { i, _ -> leftMask[i] }
        val yRight = y.filterIndexed { i, _ -> !leftMask[i] }
        
        val leftPred = if (yLeft.isNotEmpty()) yLeft.average() else 0.0
        val rightPred = if (yRight.isNotEmpty()) yRight.average() else 0.0
        
        return Tree(
            feature = bestFeature,
            threshold = bestThreshold,
            leftPred = leftPred,
            rightPred = rightPred
        )
    }
    
    private fun predictTree(X: Array<DoubleArray>, tree: Tree): DoubleArray {
        if (tree.prediction != null) {
            return DoubleArray(X.size) { tree.prediction }
        }
        
        return X.map { x ->
            if (x[tree.feature!!] <= tree.threshold!!) tree.leftPred!! else tree.rightPred!!
        }.toDoubleArray()
    }
    
    fun predict(X: Array<DoubleArray>): IntArray {
        val predictions = DoubleArray(X.size) { initialPrediction }
        
        for (tree in trees) {
            val treePredictions = predictTree(X, tree)
            for (i in predictions.indices) {
                predictions[i] += learningRate * treePredictions[i]
            }
        }
        
        return predictions.map { prediction ->
            val probability = sigmoid(prediction)
            if (probability >= 0.5) 1 else 0
        }.toIntArray()
    }
}

fun main() {
    val X = arrayOf(
        doubleArrayOf(1.0, 2.0), doubleArrayOf(2.0, 3.0), doubleArrayOf(3.0, 1.0),
        doubleArrayOf(6.0, 5.0), doubleArrayOf(7.0, 7.0), doubleArrayOf(8.0, 6.0)
    )
    val y = intArrayOf(0, 0, 0, 1, 1, 1)
    
    val model = GradientBoosting(5, 0.1, 3)
    model.fit(X, y)
    val predictions = model.predict(arrayOf(doubleArrayOf(2.0, 2.0), doubleArrayOf(7.0, 6.0)))
    println(predictions.contentToString()) // [0, 1]
}

File: gradient-boosting/gradient_boosting.py
import numpy as np

class GradientBoosting:
    def __init__(self, n_estimators=10, learning_rate=0.1, max_depth=3):
        self.n_estimators = n_estimators
        self.learning_rate = learning_rate
        self.max_depth = max_depth
        self.trees = []
    
    def sigmoid(self, x):
        return 1 / (1 + np.exp(-np.clip(x, -500, 500)))
    
    def fit(self, X, y):
        # Initialize with log odds
        self.initial_prediction = np.log(np.mean(y) / (1 - np.mean(y)))
        
        # Current predictions
        current_predictions = np.full(len(y), self.initial_prediction)
        
        for _ in range(self.n_estimators):
            # Calculate residuals (negative gradient)
            probabilities = self.sigmoid(current_predictions)
            residuals = y - probabilities
            
            # Fit a simple regression tree to residuals
            tree = self.fit_tree(X, residuals)
            self.trees.append(tree)
            
            # Update predictions
            tree_predictions = self.predict_tree(X, tree)
            current_predictions += self.learning_rate * tree_predictions
    
    def fit_tree(self, X, y):
        # Simple regression tree (just finds best split)
        best_mse = float('inf')
        best_feature, best_threshold = None, None
        
        for feature in range(X.shape[1]):
            thresholds = np.unique(X[:, feature])
            for threshold in thresholds:
                left_mask = X[:, feature] <= threshold
                if np.sum(left_mask) == 0 or np.sum(~left_mask) == 0:
                    continue
                
                y_left, y_right = y[left_mask], y[~left_mask]
                left_pred = np.mean(y_left) if len(y_left) > 0 else 0
                right_pred = np.mean(y_right) if len(y_right) > 0 else 0
                
                mse = np.sum((y_left - left_pred) ** 2) + np.sum((y_right - right_pred) ** 2)
                
                if mse < best_mse:
                    best_mse = mse
                    best_feature, best_threshold = feature, threshold
        
        if best_feature is None:
            return {'prediction': np.mean(y)}
        
        left_mask = X[:, best_feature] <= best_threshold
        left_pred = np.mean(y[left_mask]) if np.sum(left_mask) > 0 else 0
        right_pred = np.mean(y[~left_mask]) if np.sum(~left_mask) > 0 else 0
        
        return {
            'feature': best_feature,
            'threshold': best_threshold,
            'left_pred': left_pred,
            'right_pred': right_pred
        }
    
    def predict_tree(self, X, tree):
        if 'prediction' in tree:
            return np.full(len(X), tree['prediction'])
        
        predictions = np.zeros(len(X))
        for i, x in enumerate(X):
            if x[tree['feature']] <= tree['threshold']:
                predictions[i] = tree['left_pred']
            else:
                predictions[i] = tree['right_pred']
        return predictions
    
    def predict(self, X):
        predictions = np.full(len(X), self.initial_prediction)
        
        for tree in self.trees:
            tree_predictions = self.predict_tree(X, tree)
            predictions += self.learning_rate * tree_predictions
        
        probabilities = self.sigmoid(predictions)
        return (probabilities >= 0.5).astype(int)

# Test
X = np.array([[1, 2], [2, 3], [3, 1], [6, 5], [7, 7], [8, 6]])
y = np.array([0, 0, 0, 1, 1, 1])

model = GradientBoosting(n_estimators=5)
model.fit(X, y)
print(model.predict([[2, 2], [7, 6]]))  # [0, 1]
